NEON 65 - Delete by filter

jacklechnercaci opened the pull request:
Separated out delete by Id and Filter in the REST endpoints. Both use the same underlying methods other than conversion.

Added tests for all three adapters (MySQL, PostgreSQL, ES). Fixed compatibility with the new whereClause in MutateQuery.

Here are test calls: 

ElasticSearch:
```
{
  "datastoreHost": "elastic:changeme@localhost",
  "datastoreType": "elasticsearch",
  "databaseName": "earthquakes",
  "tableName": "_doc",
  "idFieldName": "testIdField",
  "dataId": "testId",
  "whereClause": {
        "type": "where",
        "lhs": {
            "database": "earthquakes",
            "table": "_doc",
            "field": "mag"
        },
        "operator": "=",
        "rhs": 1.512345678
   }
}
```

SQL:
```
{
  "datastoreHost": "user:password@localhost/earthquakes",
  "datastoreType": "postgresql",
  "databaseName": "earthquakes",
  "tableName": "quakedata",
  "idFieldName": "id",
  "dataId": "testId",
  "whereClause": {
        "type": "where",
        "lhs": {
            "database": "earthquakes",
            "table": "quakedata",
            "field": "mag"
        },
        "operator": "=",
        "rhs": 1.512345678
   }
}
```

Checked deletion of data with PgAdmin and `http://localhost:9200/earthquakes/_search?q=mag:1.512345678&pretty`

ThomasSchellenbergNextCentury commented:
@jacklechnercaci I'm getting build errors on `./gradlew clean build` in some of the test classes.  I think it's because we added a property to `MutateQuery` and so the `@AllArgsContructor` now has an additional parameter.

ThomasSchellenbergNextCentury commented on esadapter/src/main/java/com/ncc/neon/adapters/es/ElasticsearchQueryConverter.java:
```diff
@@ -430,7 +431,17 @@ public static IndexRequest convertMutationInsertQuery(MutateQuery mutateQuery) {
+        QueryBuilder queryBuilder = convertWhereClauses(selectClause, List.of(mutateQuery.getWhereClause()));
+        DeleteByQueryRequest request = new DeleteByQueryRequest();
+        request.setQuery(queryBuilder);
+        request.getSearchRequest().indices("_all");
```
I think the `indices` should be `mutateQuery.getTableName()` instead of `_all`?  See the `createSearchRequest` method in this class for a parallel use case.

jacklechnercaci replied:
it was `mutateQuery.getDatabaseName()`. Fixed.

ThomasSchellenbergNextCentury commented on server/src/main/java/com/ncc/neon/controllers/DeleteController.java:
```diff
@@ -60,4 +60,38 @@
+        }
+
+        if (invalidInput.size() > 0) {
+            return ResponseEntity.badRequest().body(Mono.just(new ActionResult("Mutation by ID Query Missing " +
```
This should say "Mutation by Filter Query Missing"

jacklechnercaci replied:
Fixed.

ThomasSchellenbergNextCentury commented on server/src/main/java/com/ncc/neon/controllers/DeleteController.java:
```diff
@@ -60,4 +60,38 @@
+    @PostMapping(path="/byfilter", consumes = MediaType.APPLICATION_JSON_UTF8_VALUE, produces = MediaType.APPLICATION_JSON_UTF8_VALUE)
+    public ResponseEntity<Mono<ActionResult>> deleteDataByFilter(@RequestBody MutateQuery mutateQuery) {
+
+        final Predicate<String> isBlank = StringUtils::isBlank;
+
+        Triple[] labeledInput = new Triple[]{
+                Triple.of("Datastore Host", mutateQuery.getDatastoreHost(), isBlank),
+                Triple.of("Datastore Type", mutateQuery.getDatastoreType(), isBlank),
+                Triple.of("Database Name", mutateQuery.getDatabaseName(), isBlank),
+                Triple.of("Table Name", mutateQuery.getTableName(), isBlank),
+                Triple.of("ID Field", mutateQuery.getIdFieldName(), isBlank)
+        };
+
+        List<String> invalidInput = Arrays.asList(labeledInput).stream()
+                .filter(triple -> ((Predicate)triple.getRight()).test(triple.getMiddle()))
+                .map(triple -> (String)triple.getLeft()).collect(Collectors.toList());
+
+        if (mutateQuery.getWhereClause() == null) {
+            invalidInput.add("Where Clause");
+        }
+
+        if (invalidInput.size() > 0) {
+            return ResponseEntity.badRequest().body(Mono.just(new ActionResult("Mutation by ID Query Missing " +
+                    String.join(", ", invalidInput))));
+        }
+
+        ConnectionInfo info = new ConnectionInfo(mutateQuery.getDatastoreType(), mutateQuery.getDatastoreHost());
+
+        datasetService.notify(new DataNotification(mutateQuery.getDatastoreHost(), mutateQuery.getDatastoreType(),
+                mutateQuery.getDatabaseName(), mutateQuery.getTableName(), 1));
+
+        return ResponseEntity.ok().body(queryService.deleteData(info, mutateQuery));
```
Could we extract most of this code into a separate method, since it's all very similar to `deleteDataById`?  I think only the `labeledInput` and the `badRequest()` string are different?

jacklechnercaci replied:
Fixed.

jacklechnercaci commented:
> @jacklechnercaci I'm getting build errors on `./gradlew clean build` in some of the test classes. I think it's because we added a property to `MutateQuery` and so the `@AllArgsContructor` now has an additional parameter.

I fixed this.
